### PR TITLE
Text classification JSON dataset reader

### PR DESCRIPTION
**Pull request by kernelmachine**
This PR introduces a text classification dataset reader which expects a JSON format.

### CODE REVIEW

File: allennlp/data/dataset_readers/text_classification_json.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    Parameters
+    ----------
+    token_indexers : ``Dict[str, TokenIndexer]``, optional
+        optional (default=``{"tokens": SingleIdTokenIndexer()}``)
+        We use this to define the input representation for the text.
+        See :class:`TokenIndexer`.
+    tokenizer : ``Tokenizer``, optional (default = ``{"tokens": WordTokenizer()}``)
+        Tokenizer to use to split the input text into words or other kinds of tokens.
+    segment_sentences: ``bool``, optional (default = ``False``)
+        If true, we will first segment the text into sentences using Spacy and then tokenize words.
```
Comment (nelson-liu) [inline allennlp/data/dataset_readers/text_classification_json.py]: `s/true/True/` and `s/Spacy/SpaCy`

File: allennlp/data/dataset_readers/text_classification_json.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        truncate a set of tokens using the provided sequence length
+        """
+        if len(tokens) > self._sequence_length:
+            tokens = tokens[:self._sequence_length]
+        return tokens
+
+    @overrides
+    def text_to_instance(self, text: str, label: str = None) -> Instance:  # type: ignore
+        """
+
```
Comment (nelson-liu) [inline allennlp/data/dataset_readers/text_classification_json.py]: extra newline

File: allennlp/data/dataset_readers/text_classification_json.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            tokens : ``TextField``
+                The tokens in the sentence or phrase.
+            label : ``LabelField``
+                The label label of the sentence or phrase.
+        """
+        # pylint: disable=arguments-differ
+        fields: Dict[str, Field] = {}
+        if self._segment_sentences:
+            sentences: List[Field] = []
+            sentence_tokens = self._sentence_segmenter.split_sentences(text)
```
Comment (nelson-liu) [inline allennlp/data/dataset_readers/text_classification_json.py]: nit: not sure if `sentence_tokens` is the right name here

File: allennlp/tests/data/dataset_readers/text_classification_json_test.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+from allennlp.data.dataset_readers import TextClassificationJSONReader
+from allennlp.common.util import ensure_list
+from allennlp.common.testing import AllenNlpTestCase
+
+
+class TestTextClassificationJSONReader():
+
+    @pytest.mark.parametrize("lazy", (True, False))
+    def test_read_from_file_ag(self, lazy):
```
Comment (nelson-liu) [inline allennlp/tests/data/dataset_readers/text_classification_json_test.py]: what does `ag` mean?

Comment (nelson-liu) [reply allennlp/tests/data/dataset_readers/text_classification_json_test.py]: oh i see, it's a dataset.

**Comment by kernelmachine** [comment]
@nelson-liu thanks for the catch. This dataset reader is meant to be a general reader for JSON files with lines of format {"text": ... , "label": ... }

File: allennlp/tests/data/dataset_readers/text_classification_json_test.py (hunk 4 above)
Comment (kernelmachine) [reply allennlp/tests/data/dataset_readers/text_classification_json_test.py]: yeah, it's a sample of the ag-news corpus. i'll make that clear.

File: allennlp/data/dataset_readers/text_classification_json.py (hunk 3 above)
Comment (kernelmachine) [reply allennlp/data/dataset_readers/text_classification_json.py]: `sentence_splits`?  `sentence_segmentations`?

File: allennlp/data/dataset_readers/text_classification_json.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+from allennlp.data.instance import Instance
+from allennlp.data.token_indexers import TokenIndexer, SingleIdTokenIndexer
+from allennlp.data.tokenizers import Tokenizer, WordTokenizer
+from allennlp.data.tokenizers.sentence_splitter import SpacySentenceSplitter
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+@DatasetReader.register("text_classification_json")
+class TextClassificationJSONReader(DatasetReader):
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/text_classification_json.py]: Nit: it should be `Json`, not `JSON`(see [here](https://google.github.io/styleguide/javaguide.html#s5.3-camel-case)).

**Comment by kernelmachine** [comment]
@nelson-liu @matt-gardner addressed all comments, ready for additional review.

**Comment by matt-gardner** [comment]
I'm fine with @nelson-liu taking this one, I was just pointing out a minor naming issue I saw from earlier.

File: allennlp/data/dataset_readers/text_classification_json.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    token_indexers : ``Dict[str, TokenIndexer]``, optional
+        optional (default=``{"tokens": SingleIdTokenIndexer()}``)
+        We use this to define the input representation for the text.
+        See :class:`TokenIndexer`.
+    tokenizer : ``Tokenizer``, optional (default = ``{"tokens": WordTokenizer()}``)
+        Tokenizer to use to split the input text into words or other kinds of tokens.
+    segment_sentences: ``bool``, optional (default = ``False``)
+        If True, we will first segment the text into sentences using SpaCy and then tokenize words.
+        Necessary for some models that require pre-segmentation of sentences,
+        like the Hierarchical Attention Network.
```
Comment (nelson-liu) [inline allennlp/data/dataset_readers/text_classification_json.py]: maybe add an arxiv link? not important.

File: allennlp/data/dataset_readers/text_classification_json.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    sequence_length: ``int``, optional (default = ``None``)
```
Comment (nelson-liu) [inline allennlp/data/dataset_readers/text_classification_json.py]: maybe use `max_sequence_length` instead? I think this makes it consistent with other pieces of code in the library.

**Pull request merged**